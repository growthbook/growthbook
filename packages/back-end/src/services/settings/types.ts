import { DataSourceInterface } from "../../../types/datasource";
import {
  ExperimentInterface,
  AttributionModel,
} from "../../../types/experiment";
import { MetricInterface } from "../../../types/metric";
import {
  OrganizationSettings,
  NorthStarMetric,
  MetricDefaults,
  ExperimentUpdateSchedule,
  MemberRoleInfo,
} from "../../../types/organization";
import { StatsEngine } from "../../../types/stats";
import { ProjectInterface } from "../../../types/project";
import { ReportInterface } from "../../../types/report";

interface SettingMetadata {
  reason?: string;
  warning?: string;
}

interface Setting<T> {
  value: T;
  meta: SettingMetadata;
}

export interface SettingsContext {
  baseSettings: ScopedSettings;
  scopes?: ScopeDefinition;
}

export type SettingsResolver<T> = (ctx: SettingsContext) => Setting<T>;

export interface ScopeDefinition {
  project?: ProjectInterface;
  datasource?: DataSourceInterface;
  experiment?: ExperimentInterface;
  metric?: MetricInterface;
  denominatorMetric?: MetricInterface;
  report?: ReportInterface;
}

export type ScopeSettingsFn = (
  scopes: ScopeDefinition
) => {
  settings: ScopedSettings;
  scopeSettings: ScopeSettingsFn;
};

interface MetricSettings {
  conversionWindowHours: number | null;
  conversionDelayHours: number | null;
  winRisk: number | null;
  loseRisk: number | null;
}

interface BaseSettings {
  confidenceLevel: number;
  northStar: NorthStarMetric | null;
  metricDefaults: MetricDefaults;
  pastExperimentsMinLength: number;
  metricAnalysisDays: number;
  updateSchedule: ExperimentUpdateSchedule | null;
  sdkInstructionsViewed: boolean;
  videoInstructionsViewed: boolean;
  multipleExposureMinPercent: number;
  defaultRole: MemberRoleInfo;
  statsEngine: StatsEngine;
  pValueThreshold: number;
  regressionAdjustmentEnabled: boolean;
  regressionAdjustmentDays: number;
  attributionModel: AttributionModel;
}

export type Settings = BaseSettings & MetricSettings;

export type ScopedSettings = Record<
  keyof Settings,
  Setting<Settings[keyof Settings]>
>;

export interface UseScopedSettingsReturn {
  settings: ScopedSettings;
  scopeSettings: ScopeSettingsFn;
}

export type InputSettings = Partial<OrganizationSettings & Settings>;
