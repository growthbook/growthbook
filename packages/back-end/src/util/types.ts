// Check if two types are equal
export type IfEqual<U, V, True, False = never> = [U] extends [V]
  ? [V] extends [U]
    ? True
    : False
  : False;
