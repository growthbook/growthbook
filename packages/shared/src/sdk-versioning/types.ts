export type SDKCapability =
  | "looseUnmarshalling"
  | "encryption"
  | "streaming"
  | "bucketingV2"
  | "visualEditor"
  | "semverTargeting"
  | "visualEditorJS"
  | "remoteEval"
  | "visualEditorDragDrop"
  | "stickyBucketing"
  | "redirects"
  | "prerequisites"
  | "savedGroupReferences";

export type CapabilityStrategy =
  | "min-ver-intersection" // intersection of capabilities using default SDK versions
  | "min-ver-intersection-loose-unmarshalling" // for generating SDK payloads. will interpret "looseUnmarshalling" as also supporting "bucketingV2", etc
  | "max-ver-intersection"; // intersection of capabilities using latest SDK versions
