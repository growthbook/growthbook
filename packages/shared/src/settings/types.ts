import { DataSourceInterface } from "back-end/types/datasource";
import {
  ExperimentInterface,
  AttributionModel,
  ExperimentInterfaceStringDates,
} from "back-end/types/experiment";
import { MetricInterface } from "back-end/types/metric";
import {
  OrganizationSettings,
  NorthStarMetric,
  MetricDefaults,
  ExperimentUpdateSchedule,
  MemberRoleInfo,
  OrganizationInterface,
  RequireReview,
} from "back-end/types/organization";
import { StatsEngine } from "back-end/types/stats";
import { ProjectInterface } from "back-end/types/project";
import { ReportInterface } from "back-end/types/report";
import { MetricWindowSettings } from "back-end/types/fact-table";

interface SettingMetadata {
  scopeApplied?: keyof ScopeDefinition | "organization";
  reason?: string;
  warning?: string;
}

interface Setting<T> {
  value: T;
  meta: SettingMetadata;
}

export interface SettingsContext {
  baseSettings: ScopedSettings;
  scopes?: ScopeDefinition;
}

export type SettingsResolver<T> = (ctx: SettingsContext) => Setting<T>;

export interface ScopeDefinition {
  organization: Partial<OrganizationInterface>;
  project?: ProjectInterface;
  datasource?: DataSourceInterface;
  experiment?: ExperimentInterface | ExperimentInterfaceStringDates;
  metric?: MetricInterface;
  denominatorMetric?: MetricInterface;
  report?: ReportInterface;
}

export type ScopeSettingsFn = (
  scopes: ScopeDefinition
) => {
  settings: ScopedSettings;
  scopeSettings: ScopeSettingsFn;
};

interface MetricSettings {
  windowType: MetricWindowSettings["type"] | null;
  windowHours: number | null;
  delayHours: number | null;

  winRisk: number | null;
  loseRisk: number | null;
}

interface BaseSettings {
  confidenceLevel: number;
  northStar: NorthStarMetric | null;
  metricDefaults: MetricDefaults;
  pastExperimentsMinLength: number;
  metricAnalysisDays: number;
  updateSchedule: ExperimentUpdateSchedule | null;
  sdkInstructionsViewed: boolean;
  videoInstructionsViewed: boolean;
  multipleExposureMinPercent: number;
  defaultRole: MemberRoleInfo;
  statsEngine: StatsEngine;
  pValueThreshold: number;
  regressionAdjustmentEnabled: boolean;
  regressionAdjustmentDays: number;
  attributionModel: AttributionModel;
  secureAttributeSalt: string;
  killswitchConfirmation: boolean;
  requireReviews: boolean | RequireReview[];
  featureKeyExample: string;
  featureRegexValidator: string;
}

// todo: encapsulate all settings, including experiment
export type Settings = BaseSettings & MetricSettings;

// export type ScopedSettings = Record<
//   keyof Settings,
//   Setting<Settings[keyof Settings]>
// >;

export type ScopedSettings = {
  [K in keyof Settings]: Setting<Settings[K]>;
};

export interface ScopedSettingsReturn {
  settings: ScopedSettings;
  scopeSettings: ScopeSettingsFn;
}

export type InputSettings = Partial<OrganizationSettings & Settings>;
