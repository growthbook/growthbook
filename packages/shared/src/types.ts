// The data going out in an sdk payload to map from a saved group ID to its array of values
export type SavedGroupsValues = Record<string, (string | number)[]>;

export type GroupMap = Map<
  string,
  Pick<SavedGroupInterface, "type" | "condition" | "attributeKey"> & {
    values?: (string | number)[];
  }
>;

export interface SavedGroupInterface {
  id: string;
  organization: string;
  groupName: string;
  owner: string;
  type: SavedGroupType;
  condition?: string;
  attributeKey?: string;
  values?: string[];
  dateUpdated: Date;
  dateCreated: Date;
  description?: string;
  projects?: string[];
}
export type SavedGroupType = "condition" | "list";
