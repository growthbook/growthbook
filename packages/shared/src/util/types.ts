/* eslint-disable @typescript-eslint/no-explicit-any */

// Check if two types are equal
export type IfEqual<U, V, True, False = never> = [U] extends [V]
  ? [V] extends [U]
    ? True
    : False
  : False;

export type IsTuple<Tuple extends any[]> = {
  empty: true;
  nonEmpty: ((...p: Tuple) => any) extends (
    p1: infer First,
    ...p: infer Rest
  ) => any
    ? IsTuple<Rest>
    : false;
  infinite: false;
}[Tuple extends []
  ? "empty"
  : Tuple extends (infer Element)[]
  ? Element[] extends Tuple
    ? "infinite"
    : "nonEmpty"
  : never];

export type TupleToUnion<T> = T extends (infer E)[] ? E : never;

export type UnionPop<U> = (
  (U extends any ? (k: (x: U) => void) => void : never) extends (
    k: infer I
  ) => void
    ? I
    : never
) extends { (a: infer A): void }
  ? A
  : never;

export type TuplePrepend<T extends any[], E> = ((
  a: E,
  ...r: T
) => void) extends (...r: infer R) => void
  ? R
  : never;

type UnionToTupleRecursively<Union, Result extends any[]> = {
  1: Result;
  0: UnionToTupleRecursively<
    Exclude<Union, UnionPop<Union>>,
    TuplePrepend<Result, UnionPop<Union>>
  >;
}[[Union] extends [never] ? 1 : 0];

export type UnionToTuple<U> = UnionToTupleRecursively<U, []>;

export function ensure<T>(x: T): asserts x is NonNullable<T> {
  if (x === undefined || x === null) {
    throw new TypeError("Internal error value should not be undefined");
  }
}

export function ensureAndReturn<T>(x: T): NonNullable<T> {
  ensure(x);
  return x;
}

// The built-in Omit<> doesn't work for certain composites like discriminated unions
export type DistributiveOmit<T, K extends PropertyKey> = T extends any
  ? Omit<T, K>
  : never;

export function isStringArray(data: unknown): data is Array<string> {
  return Array.isArray(data) && !data.find((el) => typeof el !== "string");
}

export function isString(data: unknown): data is string {
  return typeof data === "string";
}

export function isNumber(data: unknown): data is number {
  return typeof data === "number";
}
