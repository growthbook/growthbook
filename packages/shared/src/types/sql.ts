import { format as sqlFormat, FormatOptions } from "sql-formatter";

// SQL formatter dialect type that automatically stays in sync with sql-formatter
export type FormatDialect = FormatOptions["language"] | "";

export interface FormatError {
  error: Error;
  originalSql: string;
}

export function format(
  sql: string,
  dialect?: FormatDialect,
  onError?: (error: FormatError) => void
) {
  if (!dialect) return sql;

  try {
    return sqlFormat(sql, {
      language: dialect,
    });
  } catch (e) {
    const error = e instanceof Error ? e : new Error(String(e));
    if (onError) {
      onError({ error, originalSql: sql });
    } else {
      return sql;
    }
  }
}
