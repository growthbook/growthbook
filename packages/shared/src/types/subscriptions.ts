export interface PaymentMethod {
  id: string;
  brand: string;
  type: "card" | "us_bank_account" | "unknown";
  isDefault: boolean;
  last4?: string;
  expMonth?: number;
  expYear?: number;
  wallet?: string;
}
