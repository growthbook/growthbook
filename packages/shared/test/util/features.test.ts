import { FeatureInterface, FeatureRule } from "back-end/types/feature";
import { FeatureRevisionInterface } from "back-end/types/feature-revision";
import {
  OrganizationSettings,
  RequireReview,
} from "back-end/types/organization";
import {
  validateFeatureValue,
  getValidation,
  validateJSONFeatureValue,
  autoMerge,
  RulesAndValues,
  MergeConflict,
  validateCondition,
  checkEnvironmentsMatch,
  checkIfRevisionNeedsReview,
  resetReviewOnChange,
} from "../../src/util";

const feature: FeatureInterface = {
  dateCreated: new Date("2020-04-20"),
  dateUpdated: new Date("2020-04-20"),
  defaultValue: "true",
  environmentSettings: {},
  id: "feature-123",
  organization: "123",
  owner: "adnan",
  valueType: "boolean",
  version: 1,
};

const exampleJsonSchema = {
  type: "object",
  properties: {
    test: {
      type: "string",
    },
  },
};
const rules: Record<string, FeatureRule[]> = {
  dev: [
    {
      description: "test",
      id: "test",
      type: "rollout",
      value: "test",
      coverage: 1,
      hashAttribute: "test",
    },
  ],
  prod: [
    {
      description: "test",
      id: "test",
      type: "rollout",
      value: "test",
      coverage: 1,
      hashAttribute: "test",
    },
  ],
};
const changedRules: Record<string, FeatureRule[]> = {
  ...rules,
  prod: [
    ...rules.prod,

    {
      description: "test1",
      id: "test1",
      type: "rollout",
      value: "test",
      coverage: 1,
      hashAttribute: "test",
    },
  ],
};
const baseRevision: FeatureRevisionInterface = {
  featureId: feature.id,
  organization: feature.organization,
  baseVersion: 0,
  version: 0,
  dateCreated: new Date(),
  dateUpdated: new Date(),
  datePublished: null,
  publishedBy: null,
  createdBy: null,
  comment: "",
  status: "draft",
  defaultValue: "",
  rules: rules,
};

const revision: FeatureRevisionInterface = {
  featureId: feature.id,
  organization: feature.organization,
  baseVersion: 0,
  version: 1,
  dateCreated: new Date(),
  dateUpdated: new Date(),
  datePublished: null,
  publishedBy: null,
  createdBy: null,
  comment: "",
  status: "draft",
  defaultValue: "",
  rules: changedRules,
};

describe("autoMerge", () => {
  it("Auto merges when there are no conflicts", () => {
    const base: RulesAndValues = {
      defaultValue: "base",
      rules: {
        dev: [],
        prod: [],
      },
      version: 4,
    };
    const live: RulesAndValues = {
      defaultValue: "base",
      rules: {
        dev: [],
        prod: [
          {
            type: "force",
            description: "",
            id: "liveForce",
            value: "force",
          },
        ],
      },
      version: 6,
    };
    const revision: RulesAndValues = {
      defaultValue: "revision",
      rules: {
        dev: [
          {
            type: "force",
            description: "",
            id: "revisionForce",
            value: "force",
          },
        ],
        prod: [],
      },
      version: 5,
    };

    expect(autoMerge(live, base, revision, ["dev", "prod"], {})).toEqual({
      success: true,
      conflicts: [],
      result: {
        defaultValue: revision.defaultValue,
        rules: {
          dev: revision.rules["dev"],
        },
      },
    });
  });
  it("Auto merges when live and base are the same revision", () => {
    const base: RulesAndValues = {
      defaultValue: "base",
      rules: {
        dev: [],
        prod: [],
      },
      version: 4,
    };
    const revision: RulesAndValues = {
      defaultValue: "revision",
      rules: {
        dev: [
          {
            type: "force",
            description: "",
            id: "revisionForce",
            value: "force",
          },
        ],
      },
      version: 5,
    };

    expect(autoMerge(base, base, revision, ["dev", "prod"], {})).toEqual({
      success: true,
      conflicts: [],
      result: {
        defaultValue: revision.defaultValue,
        rules: {
          dev: revision.rules["dev"],
        },
      },
    });
  });
  it("Handles merge conflicts", () => {
    const base: RulesAndValues = {
      defaultValue: "base",
      rules: {
        dev: [],
        prod: [],
      },
      version: 4,
    };
    const live: RulesAndValues = {
      defaultValue: "live",
      rules: {
        dev: [],
        prod: [
          {
            type: "force",
            description: "",
            id: "liveForce",
            value: "force",
          },
        ],
      },
      version: 6,
    };
    const revision: RulesAndValues = {
      defaultValue: "revision",
      rules: {
        dev: [
          {
            type: "force",
            description: "",
            id: "revisionForce",
            value: "force",
          },
        ],
        prod: [
          {
            type: "force",
            description: "",
            id: "revisionForce",
            value: "force",
          },
        ],
      },
      version: 5,
    };

    const defaultValueConflict: MergeConflict = {
      key: "defaultValue",
      name: "Default Value",
      resolved: false,
      base: "base",
      live: "live",
      revision: "revision",
    };
    const prodConflict: MergeConflict = {
      key: "rules.prod",
      name: "Rules - prod",
      resolved: false,
      base: JSON.stringify(base.rules["prod"], null, 2),
      live: JSON.stringify(live.rules["prod"], null, 2),
      revision: JSON.stringify(revision.rules["prod"], null, 2),
    };

    expect(autoMerge(live, base, revision, ["dev", "prod"], {})).toEqual({
      success: false,
      conflicts: [defaultValueConflict, prodConflict],
    });

    expect(
      autoMerge(live, base, revision, ["dev", "prod"], {
        "rules.prod": "discard",
      })
    ).toEqual({
      success: false,
      conflicts: [
        {
          ...defaultValueConflict,
        },
        {
          ...prodConflict,
          resolved: true,
        },
      ],
    });

    expect(
      autoMerge(live, base, revision, ["dev", "prod"], {
        "rules.prod": "discard",
        defaultValue: "discard",
      })
    ).toEqual({
      success: true,
      conflicts: [
        {
          ...defaultValueConflict,
          resolved: true,
        },
        {
          ...prodConflict,
          resolved: true,
        },
      ],
      result: {
        rules: {
          dev: revision.rules["dev"],
        },
      },
    });

    expect(
      autoMerge(live, base, revision, ["dev", "prod"], {
        "rules.prod": "discard",
        defaultValue: "overwrite",
      })
    ).toEqual({
      success: true,
      conflicts: [
        {
          ...defaultValueConflict,
          resolved: true,
        },
        {
          ...prodConflict,
          resolved: true,
        },
      ],
      result: {
        defaultValue: revision.defaultValue,
        rules: {
          dev: revision.rules["dev"],
        },
      },
    });

    expect(
      autoMerge(live, base, revision, ["dev", "prod"], {
        "rules.prod": "overwrite",
        defaultValue: "overwrite",
      })
    ).toEqual({
      success: true,
      conflicts: [
        {
          ...defaultValueConflict,
          resolved: true,
        },
        {
          ...prodConflict,
          resolved: true,
        },
      ],
      result: {
        defaultValue: revision.defaultValue,
        rules: {
          dev: revision.rules["dev"],
          prod: revision.rules["prod"],
        },
      },
    });
  });
});

describe("getValidation", () => {
  it("returns validationEnabled as true if jsonSchema is populated and enabled", () => {
    feature.jsonSchema = {
      schema: JSON.stringify(exampleJsonSchema),
      date: new Date("2020-04-20"),
      enabled: true,
    };
    expect(getValidation(feature).validationEnabled).toEqual(true);
  });
  it("returns validationEnabled as false if jsonSchema enabled value is false", () => {
    feature.jsonSchema = {
      schema: JSON.stringify(exampleJsonSchema),
      date: new Date("2020-04-20"),
      enabled: false,
    };
    expect(getValidation(feature).validationEnabled).toEqual(false);
  });
  it("returns validationEnabled as false if jsonSchema is invalid", () => {
    feature.jsonSchema = {
      schema: "blahblah",
      date: new Date("2020-04-20"),
      enabled: false,
    };
    expect(getValidation(feature).validationEnabled).toEqual(false);
  });
  it("returns validationEnabled as false if jsonSchema is undefined", () => {
    feature.jsonSchema = undefined;
    expect(getValidation(feature).validationEnabled).toEqual(false);
  });
});

describe("validateJSONFeatureValue", () => {
  it("returns valid as true if all values are valid and json schema test passes", () => {
    const value = { test: "123" };
    feature.jsonSchema = {
      schema: JSON.stringify(exampleJsonSchema),
      date: new Date("2020-04-20"),
      enabled: true,
    };
    expect(validateJSONFeatureValue(value, feature).valid).toEqual(true);
  });
  it("returns valid as false if all values are valid but json schema test fails", () => {
    const value = { test: 999 };
    feature.jsonSchema = {
      schema: JSON.stringify(exampleJsonSchema),
      date: new Date("2020-04-20"),
      enabled: true,
    };
    expect(validateJSONFeatureValue(value, feature).valid).toEqual(false);
  });
  it("returns valid as false if json schema is invalid", () => {
    const value = { test: 999 };
    feature.jsonSchema = {
      schema: '{ "type": 123 }',
      date: new Date("2020-04-20"),
      enabled: true,
    };
    expect(validateJSONFeatureValue(value, feature).valid).toEqual(false);
  });
  it("returns valid as false if unparseable json value is supplied", () => {
    const value = "{ not json }";
    feature.jsonSchema = {
      schema: JSON.stringify(exampleJsonSchema),
      date: new Date("2020-04-20"),
      enabled: true,
    };
    expect(validateJSONFeatureValue(value, feature).valid).toEqual(false);
  });
  it("returns valid as true if validation is not enabled", () => {
    const value = { test: "123" };
    feature.jsonSchema = {
      schema: JSON.stringify(exampleJsonSchema),
      date: new Date("2020-04-20"),
      enabled: false,
    };
    expect(validateJSONFeatureValue(value, feature).valid).toEqual(true);
  });
});

describe("validateFeatureValue", () => {
  beforeAll(() => {
    feature.valueType = "boolean";
  });
  describe("boolean values", () => {
    it('returns "true" if value is truthy', () => {
      expect(validateFeatureValue(feature, "true", "testVal")).toEqual("true");
      expect(validateFeatureValue(feature, "0", "testVal")).toEqual("true");
    });
    it('returns "false" if value is "false"', () => {
      expect(validateFeatureValue(feature, "false", "testVal")).toEqual(
        "false"
      );
    });
  });

  describe("number values", () => {
    beforeAll(() => {
      feature.valueType = "number";
    });
    it("returns value if its a valid number", () => {
      let value = "0";
      expect(validateFeatureValue(feature, value, "testVal")).toEqual(value);
      value = "9918";
      expect(validateFeatureValue(feature, value, "testVal")).toEqual(value);
    });
    it("throws an error if value is not a valid number", () => {
      const value = "not-a-number";
      expect(() =>
        validateFeatureValue(feature, value, "testVal")
      ).toThrowError();
    });
  });

  describe("json values", () => {
    beforeAll(() => {
      feature.valueType = "json";
    });

    it("parses json and returns in string format", () => {
      const value = '{ "test": 123 }';
      expect(validateFeatureValue(feature, value, "testVal")).toEqual(
        '{"test": 123}'
      );
    });

    it('parses json that is "slightly" invalid', () => {
      let value = "{ technically: 'not valid' }";
      expect(validateFeatureValue(feature, value, "testVal")).toEqual(
        '{"technically": "not valid"}'
      );
      value = "this is not jsonbruv";
      expect(validateFeatureValue(feature, value, "testVal")).toEqual(
        `"${value}"`
      );
    });

    it("throws an error with invalid json", () => {
      const value = "{ not-an-object }";
      expect(() =>
        validateFeatureValue(feature, value, "testVal")
      ).toThrowError();
    });
  });
});

describe("validateCondition", () => {
  it("returns success when condition is undefined", () => {
    expect(validateCondition(undefined)).toEqual({
      success: true,
      empty: true,
    });
  });
  it("returns success when condition is empty", () => {
    expect(validateCondition("")).toEqual({
      success: true,
      empty: true,
    });
  });
  it("returns success when condition is empty object", () => {
    expect(validateCondition("{}")).toEqual({
      success: true,
      empty: true,
    });
  });
  it("returns error when condition is completely invalid", () => {
    expect(validateCondition("{(+")).toEqual({
      success: false,
      empty: false,
      error: "Unexpected token ( in JSON at position 1",
    });
  });
  it("returns error when condition is not an object", () => {
    expect(validateCondition("123")).toEqual({
      success: false,
      empty: false,
      error: "Must be object",
    });
  });
  it("returns suggested value when condition is invalid, but able to be fixed automatically", () => {
    expect(validateCondition("{test: true}")).toEqual({
      success: false,
      empty: false,
      error: "Unexpected token t in JSON at position 1",
      suggestedValue: '{"test":true}',
    });
  });
  it("returns success when condition is valid", () => {
    expect(validateCondition('{"test": true}')).toEqual({
      success: true,
      empty: false,
    });
  });
});

describe("check enviroments match", () => {
  it("should find a environment match", () => {
    const environments = ["prod", "staging"];
    const reviewSetting = {
      requireReviewOn: true,
      resetReviewOnChange: false,
      environments: ["prod"],
      projects: [],
    };
    expect(checkEnvironmentsMatch(environments, reviewSetting)).toEqual(true);
  });

  it("should not find a environment match", () => {
    const environments = ["prod-1"];
    const reviewSetting = {
      requireReviewOn: true,
      resetReviewOnChange: false,
      environments: ["prod"],
      projects: [],
    };
    expect(checkEnvironmentsMatch(environments, reviewSetting)).toEqual(false);
  });

  it("should turn on when everything is empty", () => {
    const environments = ["prod", "staging"];
    const reviewSetting: RequireReview = {
      requireReviewOn: true,
      resetReviewOnChange: false,
      environments: [],
      projects: [],
    };
    expect(checkEnvironmentsMatch(environments, reviewSetting)).toEqual(true);
  });
});
describe("check revision needs review", () => {
  it("should require review when env matches", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["prod"],
          projects: [],
        },
      ],
    };
    expect(
      checkIfRevisionNeedsReview({
        feature,
        baseRevision,
        revision,
        allEnvironments: ["prod", "dev", "staging"],
        settings,
      })
    ).toEqual(true);
  });
  it("should not require review", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["dev"],
          projects: [],
        },
      ],
    };
    expect(
      checkIfRevisionNeedsReview({
        feature,
        baseRevision,
        revision,
        allEnvironments: ["prod", "dev", "staging"],
        settings,
      })
    ).toEqual(false);
  });

  it("should require review with multi rules", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["dev"],
          projects: ["a"],
        },
        {
          requireReviewOn: false,
          resetReviewOnChange: false,
          environments: [],
          projects: ["b"],
        },
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["prod"],
          projects: [],
        },
      ],
    };
    expect(
      checkIfRevisionNeedsReview({
        feature,
        baseRevision,
        revision,
        allEnvironments: ["prod", "dev", "staging"],
        settings,
      })
    ).toEqual(true);
  });
  it("should not require review with multi rules", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["dev"],
          projects: [],
        },
        {
          requireReviewOn: false,
          resetReviewOnChange: false,
          environments: [],
          projects: [],
        },
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["staging"],
          projects: [],
        },
      ],
    };
    expect(
      checkIfRevisionNeedsReview({
        feature,
        baseRevision,
        revision,
        allEnvironments: ["prod", "dev", "staging"],
        settings,
      })
    ).toEqual(false);
  });
  it("legacy rules", () => {
    const settings: OrganizationSettings = {
      requireReviews: true,
    };
    expect(
      checkIfRevisionNeedsReview({
        feature,
        baseRevision,
        revision,
        allEnvironments: ["prod", "dev", "staging"],
        settings,
      })
    ).toEqual(true);
    settings.requireReviews = false;
    expect(
      checkIfRevisionNeedsReview({
        feature,
        baseRevision,
        revision,
        allEnvironments: ["prod", "dev", "staging"],
        settings,
      })
    ).toEqual(false);
  });
});

describe("reset review on change", () => {
  it("require reset with single rule", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: true,
          environments: ["prod"],
          projects: [],
        },
      ],
    };
    const settingsOff: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["prod"],
          projects: [],
        },
      ],
    };
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["staging"],
        defaultValueChanged: false,
        settings,
      })
    ).toEqual(false);
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["prod"],
        defaultValueChanged: false,
        settings,
      })
    ).toEqual(true);
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["staging"],
        defaultValueChanged: false,
        settings: settingsOff,
      })
    ).toEqual(false);
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["prod"],
        defaultValueChanged: false,
        settings: settingsOff,
      })
    ).toEqual(false);
  });

  it("require reset with multiple rules", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: true,
          environments: ["prod"],
          projects: [],
        },
        {
          requireReviewOn: true,
          resetReviewOnChange: true,
          environments: [],
          projects: [],
        },
      ],
    };
    const settingsOff: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: true,
          resetReviewOnChange: false,
          environments: ["prod"],
          projects: [],
        },
        {
          requireReviewOn: true,
          resetReviewOnChange: true,
          environments: [],
          projects: [],
        },
      ],
    };
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["staging"],
        defaultValueChanged: false,
        settings,
      })
    ).toEqual(false);
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["prod"],
        defaultValueChanged: false,
        settings,
      })
    ).toEqual(true);
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["prod"],
        defaultValueChanged: false,
        settings: settingsOff,
      })
    ).toEqual(false);
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["staging"],
        defaultValueChanged: false,
        settings: settingsOff,
      })
    ).toEqual(false);
  });
  it("turn off for first project", () => {
    const settings: OrganizationSettings = {
      requireReviews: [
        {
          requireReviewOn: false,
          resetReviewOnChange: false,
          environments: [],
          projects: ["a"],
        },
        {
          requireReviewOn: true,
          resetReviewOnChange: true,
          environments: [],
          projects: [],
        },
      ],
    };
    feature.project = "a";
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["env"],
        defaultValueChanged: false,
        settings,
      })
    ).toEqual(false);
    feature.project = "b";
    expect(
      resetReviewOnChange({
        feature,
        changedEnvironments: ["staging"],
        defaultValueChanged: false,
        settings,
      })
    ).toEqual(true);
  });
});
