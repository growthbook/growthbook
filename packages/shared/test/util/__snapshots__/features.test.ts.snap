// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getValidation returns validationEnabled as false if jsonSchema enabled value is false 1`] = `
Object {
  "jsonSchema": Object {
    "properties": Object {
      "test": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "schemaDateUpdated": 2020-04-20T00:00:00.000Z,
  "validationEnabled": false,
}
`;

exports[`getValidation returns validationEnabled as false if jsonSchema is invalid 1`] = `
Object {
  "jsonSchema": null,
  "schemaDateUpdated": null,
  "validationEnabled": false,
}
`;

exports[`getValidation returns validationEnabled as false if jsonSchema is undefined 1`] = `
Object {
  "jsonSchema": null,
  "schemaDateUpdated": undefined,
  "validationEnabled": false,
}
`;

exports[`getValidation returns validationEnabled as true if jsonSchema is populated and enabled 1`] = `
Object {
  "jsonSchema": Object {
    "properties": Object {
      "test": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "schemaDateUpdated": 2020-04-20T00:00:00.000Z,
  "validationEnabled": true,
}
`;

exports[`validateJSONFeatureValue returns valid as false if all values are valid but json schema test fails 1`] = `
Object {
  "enabled": true,
  "errors": Array [
    "Invalid value for field: test/type should be string",
  ],
  "valid": false,
}
`;

exports[`validateJSONFeatureValue returns valid as false if json schema is invalid 1`] = `
Object {
  "enabled": true,
  "errors": Array [
    "schema is invalid: data.type should be equal to one of the allowed values, data.type should be array, data.type should match some schema in anyOf",
  ],
  "valid": false,
}
`;

exports[`validateJSONFeatureValue returns valid as false if unparseable json value is supplied 1`] = `
Object {
  "enabled": true,
  "errors": Array [
    "Found } that I can't handle at line 0:12",
  ],
  "valid": false,
}
`;

exports[`validateJSONFeatureValue returns valid as true if all values are valid and json schema test passes 1`] = `
Object {
  "enabled": true,
  "errors": Array [],
  "valid": true,
}
`;

exports[`validateJSONFeatureValue returns valid as true if validation is not enabled 1`] = `
Object {
  "enabled": false,
  "errors": Array [],
  "valid": true,
}
`;
